File: Leet0094_BinaryTreeInorderTraversal/soln.py
from typing import Optional
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        output = []
        self.traverseInorder(root, output)
        return output

    def traverseInorder(self, root: Optional[TreeNode], o: List[int]) -> None:
        if root is None:
            return

        self.traverseInorder(root.left, o)
        o.append(root.val)
        self.traverseInorder(root.right, o)

        return


solution = Solution()

# Define some formatting constants
TRUNC = 1000
NULL = -999
PASS_FORMAT = "\033[92m{}\033[0m"  # Green text
FAIL_FORMAT = "\033[91m{}\033[0m"  # Red text

def create_binary_tree(values: List[int]) -> Optional[TreeNode]:
    """Convert a list of integers to a binary tree.

    The input list represents a level-order traversal of the tree,
    where NULL (-999) indicates a missing node.
    """
    if not values:
        return None

    # Create the root node
    root = TreeNode(values[0])

    # Use a queue to keep track of nodes that need children
    queue = [root]
    i = 1  # Start from the second element

    # Process the rest of the values
    while queue and i < len(values):
        # Get the next node that needs children
        current = queue.pop(0)

        # Assign left child if value is not NULL
        if i < len(values) and values[i] != NULL:
            current.left = TreeNode(values[i])
            queue.append(current.left)

        i += 1

        # Assign right child if value is not NULL
        if i < len(values) and values[i] != NULL:
            current.right = TreeNode(values[i])
            queue.append(current.right)

        i += 1

    return root

def truncate_str(s: str, max_length: int = TRUNC) -> str:
    """Truncate a string to max_length characters, adding '...' if truncated."""
    if len(s) <= max_length:
        return s
    return s[:max_length-3] + "..."

def assert_solution(input_list: List[int], expected: List[int]):
    """Test solution with given input and expected output."""
    input_copy = list(input_list)
    output = solution.inorderTraversal(create_binary_tree(input_copy))
    assert_pass = output == expected

    result = "PASS" if assert_pass else "FAIL"
    formatted_result = PASS_FORMAT.format(result) if assert_pass else FAIL_FORMAT.format(result)

    print(
        f"{formatted_result}: "
        f"Input: {truncate_str(str(input_list))}, "
        f"Expected: {truncate_str(str(expected))}"
        f"{' , Got: ' + truncate_str(str(output)) if not assert_pass else ''}"
    )

    return assert_pass

# Example 1 from problem statement
assert_solution([1, NULL, 2, 3], [1, 3, 2])

# Example 2 from problem statement
assert_solution([1, 2, 3, 4, 5, NULL, 8, NULL, NULL, 6, 7, 9], [4, 2, 6, 5, 7, 1, 3, 9, 8])

# Example 3: Empty tree
assert_solution([], [])

# Example 4: Single node tree
assert_solution([1], [1])

# Left-skewed tree
assert_solution([3, 2, NULL, 1], [1, 2, 3])

# Right-skewed tree
assert_solution([1, NULL, 2, NULL, 3], [1, 2, 3])

# Perfect binary tree
assert_solution([4, 2, 6, 1, 3, 5, 7], [1, 2, 3, 4, 5, 6, 7])

# Tree with negative values
assert_solution([0, -3, 9, -10, NULL, 5], [-10, -3, 0, 5, 9])

# Tree with duplicate values
assert_solution([5, 3, 8, 3, 4, 7, 8], [3, 3, 4, 5, 7, 8, 8])

# Tree with boundary values
assert_solution([0, -100, 100], [-100, 0, 100])

# Complex tree structure
assert_solution([8, 3, 10, 1, 6, NULL, 14, NULL, NULL, 4, 7, 13], [1, 3, 4, 6, 7, 8, 10, 13, 14])

# Deep tree - testing nested structure
assert_solution([5, 4, 8, 11, NULL, 13, 4, 7, 2, NULL, NULL, NULL, 1], [7, 11, 2, 4, 5, 13, 8, 4, 1])

# Unbalanced tree with multiple levels
assert_solution([10, 5, 15, 3, 7, NULL, 20, NULL, NULL, 6, 8], [3, 5, 6, 7, 8, 10, 15, 20])

File: Leet0110_BalancedBinaryTree/soln.java
package Leet0110_BalancedBinaryTree;

import java.util.Arrays;

public class soln {

    class Solution {

        public boolean isBalanced(TreeNode root) {

            return inOrderTraversal(root);
        }

        public boolean inOrderTraversal(TreeNode root) {

            boolean isBalanced = true;

            if (root == null) {
                return true;
            }

            int leftDepth = maxDepth(root.left);
            int rightDepth = maxDepth(root.right);

            if (Math.abs(leftDepth - rightDepth) > 1) {
                return false;
            }

            isBalanced = inOrderTraversal(root.left);
            if (isBalanced == false) {
                return false;
            }

            isBalanced = inOrderTraversal(root.right);
            if (isBalanced == false) {
                return false;
            }

            return true;
        }

        public int maxDepth(TreeNode root) {

            if (root == null) {
                return 0;
            }

            return 1 + Math.max(
                    maxDepth(root.left),
                    maxDepth(root.right));
        }
    }

    private static final int MAX_STRING_LENGTH = 50;
    private static final String ANSI_RESET = "\u001B[0m";
    private static final String ANSI_GREEN = "\u001B[92m";
    private static final String ANSI_RED = "\u001B[91m";

    private static String greenText(String text) {
        return ANSI_GREEN + text + ANSI_RESET;
    }

    private static String redText(String text) {
        return ANSI_RED + text + ANSI_RESET;
    }

    private static TreeNode createBinaryTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        // Create the root node
        TreeNode root = new TreeNode(values[0]);

        // Use a queue to keep track of nodes that need children
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(root);
        int i = 1; // Start from the second element

        // Process the rest of the values
        while (!queue.isEmpty() && i < values.length) {
            // Get the next node that needs children
            TreeNode current = queue.poll();

            // Assign left child if value is not null
            if (i < values.length) {
                if (values[i] != null) {
                    current.left = new TreeNode(values[i]);
                    queue.add(current.left);
                }
                i++;
            }

            // Assign right child if value is not null
            if (i < values.length) {
                if (values[i] != null) {
                    current.right = new TreeNode(values[i]);
                    queue.add(current.right);
                }
                i++;
            }
        }

        return root;
    }

    private static String treeToString(TreeNode root) {
        if (root == null) {
            return "Empty tree";
        }

        StringBuilder result = new StringBuilder();
        // Use a queue for level-order traversal
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(root);

        int level = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            result.append("Level ").append(level + 1).append(": ");

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();

                if (node == null) {
                    result.append("null ");
                } else {
                    result.append(node.val).append(" ");

                    // Add children to queue (including nulls for visualization)
                    queue.add(node.left);
                    queue.add(node.right);
                }
            }

            result.append("\n");
            level++;

            // Check if the next level has only nulls
            boolean allNull = true;
            for (TreeNode node : queue) {
                if (node != null) {
                    allNull = false;
                    break;
                }
            }

            if (allNull) {
                break;
            }
        }

        return result.toString();
    }

    private static String truncateStr(String s, int maxLength) {
        if (s.length() <= maxLength) {
            return s;
        }
        return s.substring(0, maxLength - 3) + "...";
    }

    private static String truncateStr(String s) {
        return truncateStr(s, MAX_STRING_LENGTH);
    }

    private static boolean assertSolution(Integer[] inputList, boolean expected) {
        soln outer = new soln();
        Solution solution = outer.new Solution();

        TreeNode tree = createBinaryTree(inputList);

        boolean output = solution.isBalanced(tree);

        boolean assertPass = output == expected;

        String result = assertPass ? "PASS" : "FAIL";
        String formattedResult = assertPass ? greenText(result) : redText(result);

        System.out.println(formattedResult + ": Input: " + truncateStr(Arrays.toString(inputList)));

        if (!assertPass) {
            System.out.println("\n=== ACTUAL TREE ===");
            System.out.println(treeToString(tree));
        }

        System.out.println("==========================================");

        return assertPass;
    }

    // Helper method to build a balanced tree of a given height
    private static Integer[] buildBalancedTree(int height) {
        int size = (int) Math.pow(2, height) - 1;
        Integer[] result = new Integer[size];
        for (int i = 0; i < size; i++) {
            result[i] = i + 1; // Just fill with sequential numbers
        }
        return result;
    }

    // Helper method to build an unbalanced (left-heavy) tree
    private static Integer[] buildUnbalancedTree(int height) {
        // Creates a tree with left side fully populated and right side empty
        int fullTreeSize = (int) Math.pow(2, height) - 1;
        Integer[] result = new Integer[fullTreeSize];

        // Root node
        result[0] = 1;

        // Construct left-heavy unbalanced tree
        for (int i = 1; i < fullTreeSize; i++) {
            // Only populate left children
            if (i % 2 == 1) {
                result[i] = i + 1;
            } else {
                result[i] = null;
            }
        }

        return result;
    }

    public static void main(String[] args) {

        // Empty tree - should be balanced
        assertSolution(new Integer[] { null },
                /* expected */ true);

        // Single node tree - should be balanced
        assertSolution(new Integer[] { 1 },
                /* expected */ true);

        // Two level balanced tree
        assertSolution(new Integer[] { 3, 9, 20 },
                /* expected */ true);

        // Three level balanced tree (example 1)
        assertSolution(new Integer[] { 3, 9, 20, null, null, 15, 7 },
                /* expected */ true);

        // Unbalanced tree with height difference of 2 (example 2)
        assertSolution(new Integer[] { 1, 2, 2, 3, 3, null, null, 4, 4 },
                /* expected */ false);

        // Unbalanced tree - left heavy
        assertSolution(new Integer[] { 1, 2, null, 3 },
                /* expected */ false);

        // Unbalanced tree - right heavy
        assertSolution(new Integer[] { 1, null, 2, null, 3 },
                /* expected */ false);

        // Perfect balanced tree (all levels completely filled)
        assertSolution(new Integer[] { 1, 2, 3, 4, 5, 6, 7 },
                /* expected */ true);

        // Balanced tree with some missing leaves
        assertSolution(new Integer[] { 1, 2, 3, null, 4, null, 5 },
                /* expected */ true);

        // Tree with balanced height but right-leaning
        assertSolution(new Integer[] { 1, null, 2, null, 3, null, 4, null, 5 },
                /* expected */ false);

        // Tree with balanced height but left-leaning
        assertSolution(new Integer[] { 1, 2, null, 3, null, 4, null, 5 },
                /* expected */ false);

        // Balanced tree with max negative values
        assertSolution(new Integer[] { -10000, -10000, -10000 },
                /* expected */ true);

        // Balanced tree with max positive values
        assertSolution(new Integer[] { 10000, 10000, 10000 },
                /* expected */ true);

        // Large balanced tree at the edge of constraints (not all 5000 nodes shown)
        assertSolution(buildBalancedTree(12), // 2^12 - 1 = 4095 nodes (approx)
                /* expected */ true);

        // Large unbalanced tree at the edge of constraints
        assertSolution(buildUnbalancedTree(12), // creates a left-heavy tree with ~4000 nodes
                /* expected */ false);

        // Tree with zigzag pattern that is balanced
        assertSolution(new Integer[] { 1, 2, 3, null, 4, null, 5, null, null, 6 },
                /* expected */ false);

        // Tree that's balanced at the root but has unbalanced subtrees
        assertSolution(new Integer[] { 3, 9, 20, 4, 5, 15, 7, 6, null, null, null, null, null, null, 8 },
                /* expected */ true);

        // LeetCode fail
        assertSolution(new Integer[] { 1, 2, 2, 3, null, null, 3, 4, null, null, 4 },
                /* expected */ false);
    }
}
